### PR TITLE
chore(ecocredit/core): keeper method audit

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

- refactors MsgCancel and MsgBridge by taking out the embedded credits, moving to it's own proto struct. This was done to avoid an unnecessary loop over credits just to change their type in the `Bridge` keeper method.
- updated errors where user provided values could return a `not found` error.
- simplifies `timestamppb` conversions in `CreateBatch`
- updates `assertClassIssuer` to use naming of `key` instead of `id`.
- removes instances where `GetCreditTypeFromBatchDenom` could just be `CreditTypeTable.Get`

ref: #715 

Closes: #1171 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/server/core/create_batch.go [hunk 1]
```diff
@@ -88,8 +88,8 @@ func (k Keeper) CreateBatch(ctx context.Context, req *core.MsgCreateBatch) (*cor
 		recipient, _ := sdk.AccAddressFromBech32(issuance.Recipient)
 
 		// get the current batch balance of the recipient account
-		// Note: Get because batch balance may or may not already exist
-		// depending on the length of issuance and if recipient is the same
+		// Note: Issuance could be for an account with no prior balance,
+		// 		 so we must catch the not found case and apply a zero value balance.
```
Comment (ryanchristo) [inline x/ecocredit/server/core/create_batch.go:92]:
```suggestion
		// Note: Issuance could be for an account with no prior balance,
		// so we must catch the not found case and apply a zero value balance.
```

**Review by ryanchristo — APPROVED** [review]
Looks good to me.

I was not sure at first on whether or not we should define `Credits` separately or define within each message but if for some reason we need to change `Credits` specifically for one message, we could theoretically redefine `Credits` as needed specifically for that message and avoid breaking changes to the other uses of `Credits`.

I also don't think this would be a likely scenario and if we were to make any changes to `Credits`, we would likely need to make the same changes to the other uses of `Credits`. I think we could maybe do something similar for `BatchIssuance`, which I can do in #1161.

**Comment by ryanchristo** [comment]
CI is breaking. Looks like unexpected error messages.

**Comment by technicallyty** [comment]
looks like there was an odd edge case in `NewDecFromString` caught by the CI.. if you pass any form of "NaN" ("NAn", "nAn", etc), it actually doesn't error out 😮

**Review by ryanchristo — APPROVED** [review]

**Comment by ryanchristo** [comment]
@aleem1314 can you review when you get a chance? :pray:

**Comment by ryanchristo** [comment]
> looks like there was an odd edge case in NewDecFromString caught by the CI.. if you pass any form of "NaN" ("NAn", "nAn", etc), it actually doesn't error, but the test expects one open_mouth

Opened #1181

**Review by aleem1314 — APPROVED** [review]

**Comment by ryanchristo** [comment]
cc @technicallyty can you resolve conflicts? :pray:

File: x/ecocredit/server/core/create_batch.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ func (k Keeper) CreateBatch(ctx context.Context, req *core.MsgCreateBatch) (*cor
 		return nil, err
 	}
 
-	startDate, endDate := timestamppb.New(req.StartDate.UTC()), timestamppb.New(req.EndDate.UTC())
+	startDate, endDate := timestamppb.New(*req.StartDate), timestamppb.New(*req.EndDate)
```
Comment (ryanchristo) [inline x/ecocredit/server/core/create_batch.go:54]: I missed this the first time around... I don't think we want to store local times if the user provides a local time. I think we always want to store the UTC time.

Comment (ryanchristo) [reply x/ecocredit/server/core/create_batch.go:54]: cc @clevinson

Comment (technicallyty) [reply x/ecocredit/server/core/create_batch.go]: whoops, nice catch - fixed here https://github.com/regen-network/regen-ledger/pull/1160/commits/524e2f5cf0b7f930f05e299575a778ff7a1769d5
